# Distributed SparkML Model Workflow

## 📌 Project Overview

This notebook demonstrates the creation, training, saving, loading, and distributed inference execution of Machine Learning models using Apache Spark MLlib.

The workflow simulates a scalable SparkML pipeline where regression models are trained on structured datasets and deployed for distributed prediction processing.

The notebook focuses on:
- SparkML model training
- feature engineering
- distributed inference workflows
- model persistence
- scalable prediction execution
- Spark-based Machine Learning operations

---

## 🎯 Objectives

- Initialize Apache Spark environment
- Create Spark DataFrames for ML processing
- Transform features using VectorAssembler
- Train Linear Regression models
- Save SparkML models
- Load pretrained SparkML models
- Execute distributed predictions
- Demonstrate scalable ML inference workflows

---

## 🛠 Technologies Used

- Apache Spark
- PySpark
- Spark MLlib
- Linear Regression
- VectorAssembler
- Distributed Machine Learning
- Predictive Analytics
- Python
- Jupyter Notebook

---

## 🏗 Machine Learning Workflow

The workflow includes:
- Spark environment initialization
- structured dataset creation
- feature vector transformation
- ML model training
- model persistence and serialization
- distributed prediction execution
- inference result generation

The notebook demonstrates a simplified enterprise Machine Learning pipeline using Apache Spark MLlib.

# 1. Install Required Packages

The required PySpark libraries are installed to initialize the Apache Spark environment and enable distributed Machine Learning workflows.

The installation includes:
- PySpark runtime
- Spark environment discovery utilities
- distributed processing dependencies

These packages provide the foundation for Spark DataFrame operations and SparkML model execution.

In [1]:
!pip install pyspark
!pip install findspark

## Installation Results

The required Spark packages were successfully installed.

Installed components:
- PySpark
- findspark
- Spark Python dependencies

The environment is now prepared for distributed Spark processing and Machine Learning operations.

# 2. Initialize Spark Environment

The Apache Spark environment is initialized using `findspark`, followed by importing the required Spark modules for distributed data processing and Machine Learning operations.

The initialization process prepares:
- SparkContext
- SparkSession
- distributed execution environment
- Spark DataFrame processing capabilities

In [2]:
import findspark
findspark.init()

from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession

## Spark Environment Initialization Results

The Spark environment was successfully initialized.

Initialized components:
- SparkContext libraries
- SparkSession dependencies
- distributed Spark runtime
- Spark SQL processing modules

The notebook environment is now prepared for distributed analytics and SparkML workflows.

# 3. Create Spark Session

A Spark session is created as the main entry point for working with Spark DataFrames and Machine Learning operations.

The Spark session enables:
- distributed DataFrame processing
- Spark SQL operations
- SparkML execution
- scalable analytical workflows

In [4]:
# Create Spark session
spark = SparkSession \
    .builder \
    .appName("Distributed SparkML Model Workflow") \
    .getOrCreate()

sc = spark.sparkContext

## Spark Session Results

The Spark session was successfully created.

Session capabilities:
- distributed computation support
- Spark DataFrame processing
- SparkML integration
- scalable execution environment

The Spark runtime is now fully operational for Machine Learning and predictive analytics workflows.

# 4. Import Spark ML Libraries

Spark ML libraries are imported to enable feature engineering, regression modeling, and distributed Machine Learning workflows.

The imported libraries provide:
- feature vector transformation
- regression model training
- distributed ML processing
- predictive analytics functionality

In [5]:
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.regression import LinearRegression

## Spark ML Library Import Results

The required Spark ML libraries were successfully imported.

Imported components:
- VectorAssembler
- LinearRegression
- SparkML feature engineering tools
- distributed regression modeling libraries

The environment is now prepared for Machine Learning model development and inference workflows.

# 5. Create a Spark DataFrame with Sample Data

A structured dataset is created to simulate a Machine Learning regression workflow using Apache Spark.

The dataset contains:
- infant height measurements
- infant weight measurements
- numerical features for regression analysis

The dataset is converted into a Spark DataFrame for distributed processing and SparkML model training.

In [7]:
# Create sample dataset
mydata = [[46,2.5],[51,3.4],[54,4.4],[57,5.1],[60,5.6],[61,6.1],[63,6.4]]

# Define columns
columns = ["height", "weight"]

# Create Spark DataFrame
mydf = spark.createDataFrame(mydata, columns)

# Display dataset
mydf.show()

+------+------+
|height|weight|
+------+------+
|    46|   2.5|
|    51|   3.4|
|    54|   4.4|
|    57|   5.1|
|    60|   5.6|
|    61|   6.1|
|    63|   6.4|
+------+------+



## Dataset Preparation Results

The sample dataset was successfully converted into a Spark DataFrame for Machine Learning processing.

Dataset characteristics:
- total records: 7
- numerical regression dataset
- input feature: `height`
- target variable: `weight`
- structured tabular format
- Spark DataFrame compatibility
- suitable for regression model training

The dataset is now prepared for feature engineering and SparkML workflows.

# 6. Feature Engineering with VectorAssembler

Feature engineering is performed using the `VectorAssembler` class to transform numerical input columns into Spark ML feature vectors.

This transformation is required because SparkML algorithms operate on vectorized feature representations.

The workflow includes:
- feature column transformation
- vector creation
- ML-compatible dataset preparation
- structured feature engineering for regression modeling

In [8]:
assembler = VectorAssembler(
    inputCols=["height"],
    outputCol="features"
)

data = assembler.transform(mydf).select('features', 'weight')

data.show()

+--------+------+
|features|weight|
+--------+------+
|  [46.0]|   2.5|
|  [51.0]|   3.4|
|  [54.0]|   4.4|
|  [57.0]|   5.1|
|  [60.0]|   5.6|
|  [61.0]|   6.1|
|  [63.0]|   6.4|
+--------+------+



## Feature Engineering Results

The VectorAssembler successfully transformed the input feature columns into Spark ML feature vectors.

Key observations:
- numerical height values were successfully converted into vectorized features
- 7 records were transformed into SparkML-compatible feature vectors
- the dataset structure is compatible with distributed SparkML algorithms
- the transformed DataFrame is ready for regression model training
- feature vectors were successfully generated for distributed ML processing

Generated dataset structure:
- feature column: `features`
- target column: `weight`
- total transformed records: 7
- sample feature vector: `[46.0]`
- sample target value: `2.5`

The feature engineering stage successfully prepared the dataset for distributed Machine Learning processing.

# 7. Train the Linear Regression Model

A Linear Regression model is created and trained using Spark MLlib to simulate a distributed Machine Learning workflow.

The training process includes:
- regression model initialization
- parameter configuration
- distributed model training
- predictive relationship learning between height and weight values

The model is trained using SparkML distributed processing capabilities.

In [10]:
# Create Linear Regression model
lr = LinearRegression(featuresCol='features', labelCol='weight', maxIter=100)

lr.setRegParam(0.1)

# Train model
lrModel = lr.fit(data)

## Model Training Results

The Linear Regression model was successfully trained using Spark MLlib.

Training workflow highlights:
- distributed regression training executed successfully
- feature vectors processed correctly
- regression parameters initialized successfully
- SparkML training pipeline completed without execution errors
- predictive relationship learning completed using distributed Spark processing

Training configuration:
- algorithm: Linear Regression
- maximum iterations: 100
- regularization parameter: 0.1
- input feature: `height`
- prediction target: `weight`

The trained model is now ready for persistence and distributed inference execution.

# 8. Save the SparkML Model

The trained SparkML model is saved to persistent storage to simulate enterprise Machine Learning model lifecycle management.

Model persistence enables:
- reusable ML workflows
- distributed model deployment
- scalable inference execution
- model versioning and storage

In [20]:
lrModel.write().overwrite().save("../models/infantheight2.model")

## Model Persistence Results

The SparkML model was successfully saved to persistent storage.

Implemented operations:
- model serialization
- model persistence
- reusable ML workflow creation
- SparkML model export

Saved model details:
- model name: `infantheight2.model`
- model type: Linear Regression
- storage operation: overwrite enabled
- persistence workflow completed successfully

The trained model is now available for distributed loading and inference workflows.

# 9. Load the SparkML Model

The previously saved SparkML model is loaded from persistent storage for distributed inference execution.

The loading workflow demonstrates:
- model deserialization
- reusable inference pipelines
- SparkML deployment workflows
- distributed prediction preparation

In [21]:
from pyspark.ml.regression import LinearRegressionModel

model = LinearRegressionModel.load('../models/infantheight2.model')

## Model Loading Results

The pretrained SparkML model was successfully loaded from persistent storage.

Loaded model capabilities:
- distributed inference support
- reusable regression predictions
- SparkML deployment compatibility
- scalable predictive analytics execution

Loaded model details:
- loaded model: `infantheight2.model`
- model type: Linear Regression
- SparkML deserialization completed successfully

The model is now ready for distributed prediction execution.

# 10. Execute Distributed Predictions

The loaded SparkML model is used to generate predictions for new input values using distributed Spark inference workflows.

The prediction pipeline includes:
- feature vector generation
- DataFrame transformation
- distributed inference execution
- regression prediction generation

The workflow simulates enterprise predictive analytics execution using Apache Spark.

In [14]:
# Prediction function
def predict(height):
    assembler = VectorAssembler(
        inputCols=["height"],
        outputCol="features"
    )

    data = [[height, 0]]
    columns = ["height", "weight"]

    df = spark.createDataFrame(data, columns)

    transformed_df = assembler.transform(df).select('features', 'weight')

    predictions = model.transform(transformed_df)

    predictions.select("prediction").show()

# 11. Predict Infant Weight for Height = 70 cm

A distributed prediction is executed using the pretrained SparkML regression model.

The model generates an estimated weight prediction based on the provided height input value.

In [15]:
predict(70)

+-----------------+
|       prediction|
+-----------------+
|7.863454719775907|
+-----------------+



## Prediction Results

The pretrained SparkML model successfully generated prediction outputs for new input values.

Prediction observations:
- distributed inference executed successfully
- feature transformation pipeline worked correctly
- regression prediction values were generated
- SparkML inference workflow completed successfully

Prediction summary:
- input height: `70 cm`
- predicted weight: `7.86 kg`
- raw prediction value: `7.863454719775907`
- prediction engine: Spark MLlib Linear Regression
- inference workflow: distributed Spark execution

The notebook demonstrates scalable predictive analytics using Apache Spark MLlib.

# 12. Save the Model as `babyweightprediction.model`

The trained SparkML regression model is saved again using a different model name to demonstrate reusable model persistence workflows and model versioning practices.

This workflow simulates:
- enterprise model management
- reusable ML deployments
- model version persistence
- scalable SparkML lifecycle operations

In [22]:
lrModel.write().overwrite().save('../models/babyweightprediction.model')

## Secondary Model Persistence Results

The SparkML model was successfully saved using a secondary model identifier.

Persistence workflow highlights:
- model overwrite operation completed successfully
- reusable model deployment enabled
- SparkML serialization executed correctly
- model lifecycle workflow validated

Saved model details:
- model name: `babyweightprediction.model`
- model type: Linear Regression
- persistence operation: successful

The secondary model is now available for distributed inference workflows.

# 13. Load the `babyweightprediction.model`

The previously saved regression model is loaded from persistent storage to validate reusable SparkML deployment and inference workflows.

The loading operation demonstrates:
- model deserialization
- reusable SparkML pipelines
- scalable predictive analytics workflows
- distributed inference preparation

In [23]:
model = LinearRegressionModel.load('../models/babyweightprediction.model')

## Secondary Model Loading Results

The `babyweightprediction.model` was successfully loaded into the Spark environment.

Loaded model capabilities:
- distributed inference execution
- reusable regression predictions
- scalable SparkML workflows
- predictive analytics support

Loaded model details:
- loaded model: `babyweightprediction.model`
- model type: Linear Regression
- loading operation completed successfully

The model is now ready for additional prediction workflows.

# 14. Predict Infant Weight for Height = 50 cm

The loaded SparkML regression model is used to generate a prediction for a new infant height input value.

This workflow validates:
- reusable prediction pipelines
- distributed Spark inference execution
- scalable ML prediction generation

In [19]:
predict(50)

+------------------+
|        prediction|
+------------------+
|3.4666826711164465|
+------------------+



## Final Prediction Results

The SparkML regression model successfully generated prediction results for the new input value.

Prediction observations:
- distributed inference executed successfully
- SparkML prediction pipeline operated correctly
- regression output generated without execution errors
- reusable ML workflow validated successfully

Prediction summary:
- input height: `50 cm`
- predicted weight: `3.47 kg`
- raw prediction value: `3.4666826711164465`
- prediction engine: Spark MLlib Linear Regression
- execution environment: Apache Spark distributed runtime

The notebook successfully demonstrates distributed Machine Learning inference and reusable SparkML deployment workflows.

# ✅ Notebook Outcome

This notebook successfully demonstrated an end-to-end distributed Machine Learning workflow using Apache Spark MLlib.

Implemented capabilities:
- Spark environment initialization
- Spark DataFrame creation and processing
- feature engineering with VectorAssembler
- distributed Linear Regression model training
- SparkML model persistence and loading
- distributed prediction execution
- scalable Machine Learning inference workflows

Key achievements:
- structured datasets successfully processed with Spark
- feature vectors generated for ML workflows
- regression models trained using distributed Spark execution
- reusable SparkML deployment pipelines implemented
- predictive analytics workflows validated successfully

The notebook demonstrates enterprise-style Big Data and Machine Learning engineering workflows using Apache Spark.